main.go

```diff
@@ -121,95 +124,94 @@ func run(c *cli.Context) error {
 	return runWorker(c, n, gateway)
 }
 
-func runGateway(c *cli.Context, n *server.Node) error {
-	fmt.Println("Gateway booting up...")
-	time.Sleep(5 * time.Second)
-	fmt.Println("Running Gateway...")
-
-	var ch = csp.NewChan(&csp.SyncChan{})
-	n.Vat.Export(chanCap, chanProvider{ch}) // Sets location to "/lb/chan"
-
-	var err error
-	for err == nil {
-		f, release := ch.Recv(context.Background())
-		defer release()
-
-		// TODO: uncomment to go from: Ptr => Client
-		// ptr, err := f.Ptr()
-		// if err != nil {
-		// 	return err
-		// }
-		// client := ptr.Interface().Client()
-		// a := handler.Handler(client)
-		// val := // TODO: Go from http request => Value
-		// a.Handle(context.Background(), val)
-
-		got, err := f.Text()
-
-		if err != nil {
-			return err
-		}
-		logger.Info("We have received value: " + got + " from the channel!")
-		time.Sleep(time.Second)
-	}
-	// TODO: Uncomment later when want to add in http server functionality
-	// logger.Info("starting server, listening on port :8080")
-
-	// http.HandleFunc("/echo", EchoHandler)
-	// http.HandleFunc("/slight-echo", SlightEchoHandler)
-
-	//return http.ListenAndServe(":8080", nil) // Can test with: curl -X GET -H "Content-Type: application/json" -d '{"message": "Hello, World!"}' http://localhost:8080/slight-echo
-	// time.Sleep(10 * time.Second)
-	return err
+type Payload struct {
+	Message string            `json:"message"`
+	Status  int               `json:"status"`
+	Headers map[string]string `json:"headers"`
+	Body    string            `json:"body"`
 }
 
-// EchoHandler echos back the request as a response
-// From https://github.com/aautar/go-http-echo/blob/master/echo.go#L21-L40
-func EchoHandler(writer http.ResponseWriter, request *http.Request) {
-
-	logger.Info("Echoing back request made to " + request.URL.Path + " to client (" + request.RemoteAddr + ")")
-
-	// allow pre-flight headers
-	writer.Header().Set("Access-Control-Allow-Headers", "Content-Range, Content-Disposition, Content-Type, ETag")
+func gatewayHandler(ch csp.Chan, c *cli.Context, n *server.Node, writer http.ResponseWriter, request *http.Request) {
 
-	request.Write(writer)
-}
+	// Receive value from synchronous channel
+	f, r := ch.Recv(context.Background())
+	defer r()
 
-type Payload struct {
-	Message string `json:"message"`
-}
+	// Turn received value into pointer
+	ptr, err := f.Ptr()
+	if err != nil {
+		panic(err)
+	}
 
-func SlightEchoHandler(writer http.ResponseWriter, request *http.Request) {
-	log.Info("Slightly echoing back request made to " + request.URL.Path + " to client (" + request.RemoteAddr + ")")
+	// Recover capability from client
+	a := handler.Handler(ptr.Interface().Client())
 
+	// Decode object
 	var payload Payload
-
-	err := json.NewDecoder(request.Body).Decode(&payload)
+	err = json.NewDecoder(request.Body).Decode(&payload)
```

Comment:
This is line I am worried about (with respect to marshaling / unmarshaling) @lthibault

Reply:
Ah yeah, my advice would be to skip the JSON.  I would just echo the request body back to the caller, byte-for-byte.

Steps are:

1. Use `io.ReadAll(r.Body)` to get the request body
2. Hand it off to some worker's echo capability
3. Return the result from the `echo` by writing it to the `http.ResponseWriter`.